File: JobTracker/Services/GmailAuthService.swift
//
//  GmailAuthService.swift
//  JobTracker
//
//  Dependency-free Google OAuth 2.0 for a native macOS app using PKCE and
//  ASWebAuthenticationSession. Tokens are persisted in the Keychain and
//  refreshed on demand.
//

import Foundation
import AuthenticationServices
import CryptoKit
import AppKit

enum AuthError: LocalizedError {
    case notConfigured
    case cancelled
    case badResponse(String)
    case noRefreshToken

    var errorDescription: String? {
        switch self {
        case .notConfigured:   return "Google client ID is not set. Add it in Settings or AppConfig.swift."
        case .cancelled:       return "Sign-in was cancelled."
        case .badResponse(let m): return "OAuth error: \(m)"
        case .noRefreshToken:  return "No refresh token available. Please sign in again."
        }
    }
}

@MainActor
@Observable
final class GmailAuthService: NSObject {

    var isSignedIn: Bool = Secrets.shared.get(.gmailRefreshToken) != nil
    var accountEmail: String?

    private var currentSession: ASWebAuthenticationSession?

    // MARK: - Sign in

    func signIn() async throws {
        guard AppConfig.isGoogleConfigured else { throw AuthError.notConfigured }

        let verifier = Self.randomCodeVerifier()
        let challenge = Self.codeChallenge(for: verifier)

        var components = URLComponents(url: AppConfig.authorizationEndpoint, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            .init(name: "client_id", value: AppConfig.googleClientID),
            .init(name: "redirect_uri", value: AppConfig.redirectURI),
            .init(name: "response_type", value: "code"),
            .init(name: "scope", value: AppConfig.gmailScope),
            .init(name: "code_challenge", value: challenge),
            .init(name: "code_challenge_method", value: "S256"),
            .init(name: "access_type", value: "offline"),
            .init(name: "prompt", value: "consent")
        ]

        let callbackURL = try await presentAuthSession(url: components.url!)

        guard let code = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?
            .queryItems?.first(where: { $0.name == "code" })?.value else {
            throw AuthError.badResponse("Missing authorization code")
        }

        try await exchangeCode(code, verifier: verifier)
        isSignedIn = true
    }

    private func presentAuthSession(url: URL) async throws -> URL {
        try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: url,
                callbackURLScheme: AppConfig.callbackURLScheme
            ) { callbackURL, error in
                if let callbackURL {
                    continuation.resume(returning: callbackURL)
                } else if let error = error as? ASWebAuthenticationSessionError,
                          error.code == .canceledLogin {
                    continuation.resume(throwing: AuthError.cancelled)
                } else {
                    continuation.resume(throwing: error ?? AuthError.cancelled)
                }
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            self.currentSession = session
            session.start()
        }
    }

    // MARK: - Token exchange & refresh

    private func exchangeCode(_ code: String, verifier: String) async throws {
        let params = [
            "client_id": AppConfig.googleClientID,
            "code": code,
            "code_verifier": verifier,
            "grant_type": "authorization_code",
            "redirect_uri": AppConfig.redirectURI
        ]
        let token = try await postToken(params)
        persist(token)
        if let idToken = token.idToken { accountEmail = Self.email(fromIDToken: idToken) }
    }

    /// Returns a valid access token, refreshing if necessary.
    func validAccessToken() async throws -> String {
        if let token = Secrets.shared.get(.gmailAccessToken),
           let expiryString = Secrets.shared.get(.gmailAccessTokenExpiry),
           let expiry = TimeInterval(expiryString),
           Date().timeIntervalSince1970 < expiry - 60 {
            return token
        }
        return try await refresh()
    }

    private func refresh() async throws -> String {
        guard let refreshToken = Secrets.shared.get(.gmailRefreshToken) else {
            throw AuthError.noRefreshToken
        }
        let params = [
            "client_id": AppConfig.googleClientID,
            "refresh_token": refreshToken,
            "grant_type": "refresh_token"
        ]
        let token = try await postToken(params)
        persist(token)
        guard let access = token.accessToken else { throw AuthError.badResponse("No access token") }
        return access
    }

    private func postToken(_ params: [String: String]) async throws -> TokenResponse {
        var request = URLRequest(url: AppConfig.tokenEndpoint)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = params
            .map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryValueAllowed) ?? $0.value)" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw AuthError.badResponse(String(data: data, encoding: .utf8) ?? "Unknown")
        }
        return try JSONDecoder().decode(TokenResponse.self, from: data)
    }

    private func persist(_ token: TokenResponse) {
        if let access = token.accessToken {
            Secrets.shared.set(access, for: .gmailAccessToken)
            let expiry = Date().timeIntervalSince1970 + Double(token.expiresIn ?? 3600)
            Secrets.shared.set(String(expiry), for: .gmailAccessTokenExpiry)
        }
        if let refresh = token.refreshToken {
            Secrets.shared.set(refresh, for: .gmailRefreshToken)
        }
    }

    func signOut() {
        Secrets.shared.set(nil, for: .gmailAccessToken)
        Secrets.shared.set(nil, for: .gmailRefreshToken)
        Secrets.shared.set(nil, for: .gmailAccessTokenExpiry)
        isSignedIn = false
        accountEmail = nil
    }

    // MARK: - PKCE helpers

    private static func randomCodeVerifier() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64URLEncodedString()
    }

    private static func codeChallenge(for verifier: String) -> String {
        let hash = SHA256.hash(data: Data(verifier.utf8))
        return Data(hash).base64URLEncodedString()
    }

    /// Extracts the email claim from a JWT id_token without verifying it
    /// (verification isn't needed — the token came directly from Google over TLS).
    private static func email(fromIDToken idToken: String) -> String? {
        let parts = idToken.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
        payload = payload.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while payload.count % 4 != 0 { payload += "=" }
        guard let data = Data(base64Encoded: payload),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return json["email"] as? String
    }
}

// MARK: - Presentation anchor

extension GmailAuthService: ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        NSApplication.shared.keyWindow ?? ASPresentationAnchor()
    }
}

// MARK: - Token model

private struct TokenResponse: Decodable {
    let accessToken: String?
    let refreshToken: String?
    let expiresIn: Int?
    let idToken: String?

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case idToken = "id_token"
    }
}

extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

extension CharacterSet {
    static let urlQueryValueAllowed: CharacterSet = {
        var set = CharacterSet.alphanumerics
        set.insert(charactersIn: "-._~")
        return set
    }()
}

File: JobTracker/Services/GmailAuthService.cs
// Dependency-free Google OAuth 2.0 for a native Windows app using PKCE and
// the system browser + a loopback HTTP listener (Google's documented flow
// for installed/desktop apps — the Windows analog of macOS's
// ASWebAuthenticationSession). Tokens are persisted via Secrets (DPAPI) and
// refreshed on demand.
//
// IMPORTANT for users setting up their own OAuth client: this flow needs a
// "Desktop app" type client (not "iOS", which the macOS build uses) since
// it relies on the http://127.0.0.1:{port}/ loopback redirect.

using System.Diagnostics;
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using CommunityToolkit.Mvvm.ComponentModel;

namespace JobTracker.Services;

public enum AuthErrorKind { NotConfigured, Cancelled, BadResponse, NoRefreshToken }

public sealed class AuthException(AuthErrorKind kind, string? detail = null)
    : Exception(Describe(kind, detail))
{
    public AuthErrorKind Kind { get; } = kind;

    private static string Describe(AuthErrorKind kind, string? detail) => kind switch
    {
        AuthErrorKind.NotConfigured => "Google client ID is not set. Add it in Settings.",
        AuthErrorKind.Cancelled => "Sign-in was cancelled.",
        AuthErrorKind.BadResponse => $"OAuth error: {detail}",
        AuthErrorKind.NoRefreshToken => "No refresh token available. Please sign in again.",
        _ => "OAuth error",
    };
}

public sealed partial class GmailAuthService : DispatcherObservableObject
{
    [ObservableProperty]
    private bool _isSignedIn = Secrets.Shared.Get(SecretKey.GmailRefreshToken) is not null;

    [ObservableProperty]
    private string? _accountEmail;

    private static readonly HttpClient Http = new();
    private static readonly TimeSpan CallbackTimeout = TimeSpan.FromMinutes(5);

    // MARK: - Sign in

    public async Task SignInAsync()
    {
        if (!AppConfig.IsGoogleConfigured) throw new AuthException(AuthErrorKind.NotConfigured);

        var verifier = RandomCodeVerifier();
        var challenge = CodeChallenge(verifier);
        var port = GetFreeLoopbackPort();
        var redirectUri = $"http://127.0.0.1:{port}/oauth2redirect/";

        using var listener = new HttpListener();
        listener.Prefixes.Add(redirectUri);
        listener.Start();

        var authorizationUrl = BuildAuthorizationUrl(redirectUri, challenge);
        Process.Start(new ProcessStartInfo(authorizationUrl) { UseShellExecute = true });

        string code;
        try
        {
            code = await WaitForCallbackAsync(listener);
        }
        finally
        {
            listener.Stop();
        }

        await ExchangeCodeAsync(code, verifier, redirectUri);
        IsSignedIn = true;
    }

    private static int GetFreeLoopbackPort()
    {
        var listener = new TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }

    private static async Task<string> WaitForCallbackAsync(HttpListener listener)
    {
        var contextTask = listener.GetContextAsync();
        var completed = await Task.WhenAny(contextTask, Task.Delay(CallbackTimeout));
        if (completed != contextTask)
        {
            throw new AuthException(AuthErrorKind.Cancelled, "Sign-in timed out.");
        }

        var context = await contextTask;
        var query = context.Request.QueryString;
        var code = query["code"];
        var error = query["error"];

        var html = error is null
            ? "<html><body>Signed in — you can close this window.</body></html>"
            : "<html><body>Sign-in failed. You can close this window.</body></html>";
        var buffer = Encoding.UTF8.GetBytes(html);
        context.Response.ContentType = "text/html";
        context.Response.ContentLength64 = buffer.Length;
        await context.Response.OutputStream.WriteAsync(buffer);
        context.Response.OutputStream.Close();

        if (error is not null) throw new AuthException(AuthErrorKind.BadResponse, error);
        if (string.IsNullOrEmpty(code)) throw new AuthException(AuthErrorKind.BadResponse, "Missing authorization code");
        return code;
    }

    private static string BuildAuthorizationUrl(string redirectUri, string challenge)
    {
        var query = new Dictionary<string, string>
        {
            ["client_id"] = AppConfig.GoogleClientId,
            ["redirect_uri"] = redirectUri,
            ["response_type"] = "code",
            ["scope"] = AppConfig.GmailScope,
            ["code_challenge"] = challenge,
            ["code_challenge_method"] = "S256",
            ["access_type"] = "offline",
            ["prompt"] = "consent",
        };
        var qs = string.Join('&', query.Select(kv => $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));
        return $"{AppConfig.AuthorizationEndpoint}?{qs}";
    }

    // MARK: - Token exchange & refresh

    private async Task ExchangeCodeAsync(string code, string verifier, string redirectUri)
    {
        var token = await PostTokenAsync(new Dictionary<string, string>
        {
            ["client_id"] = AppConfig.GoogleClientId,
            ["code"] = code,
            ["code_verifier"] = verifier,
            ["grant_type"] = "authorization_code",
            ["redirect_uri"] = redirectUri,
        });
        Persist(token);
        if (token.IdToken is not null) AccountEmail = EmailFromIdToken(token.IdToken);
    }

    /// Returns a valid access token, refreshing if necessary.
    public async Task<string> ValidAccessTokenAsync()
    {
        var token = Secrets.Shared.Get(SecretKey.GmailAccessToken);
        var expiryString = Secrets.Shared.Get(SecretKey.GmailAccessTokenExpiry);
        if (token is not null && expiryString is not null &&
            double.TryParse(expiryString, NumberStyles.Float, CultureInfo.InvariantCulture, out var expiry) &&
            DateTimeOffset.UtcNow.ToUnixTimeSeconds() < expiry - 60)
        {
            return token;
        }
        return await RefreshAsync();
    }

    private async Task<string> RefreshAsync()
    {
        var refreshToken = Secrets.Shared.Get(SecretKey.GmailRefreshToken)
            ?? throw new AuthException(AuthErrorKind.NoRefreshToken);
        var token = await PostTokenAsync(new Dictionary<string, string>
        {
            ["client_id"] = AppConfig.GoogleClientId,
            ["refresh_token"] = refreshToken,
            ["grant_type"] = "refresh_token",
        });
        Persist(token);
        return token.AccessToken ?? throw new AuthException(AuthErrorKind.BadResponse, "No access token");
    }

    private static async Task<TokenResponse> PostTokenAsync(Dictionary<string, string> parameters)
    {
        using var content = new FormUrlEncodedContent(parameters);
        using var response = await Http.PostAsync(AppConfig.TokenEndpoint, content);
        var body = await response.Content.ReadAsStringAsync();
        if (!response.IsSuccessStatusCode) throw new AuthException(AuthErrorKind.BadResponse, body);
        return JsonSerializer.Deserialize<TokenResponse>(body)
            ?? throw new AuthException(AuthErrorKind.BadResponse, "Empty response");
    }

    private static void Persist(TokenResponse token)
    {
        if (token.AccessToken is not null)
        {
            Secrets.Shared.Set(token.AccessToken, SecretKey.GmailAccessToken);
            var expiry = DateTimeOffset.UtcNow.ToUnixTimeSeconds() + (token.ExpiresIn ?? 3600);
            Secrets.Shared.Set(expiry.ToString(CultureInfo.InvariantCulture), SecretKey.GmailAccessTokenExpiry);
        }
        if (token.RefreshToken is not null)
        {
            Secrets.Shared.Set(token.RefreshToken, SecretKey.GmailRefreshToken);
        }
    }

    public void SignOut()
    {
        Secrets.Shared.Set(null, SecretKey.GmailAccessToken);
        Secrets.Shared.Set(null, SecretKey.GmailRefreshToken);
        Secrets.Shared.Set(null, SecretKey.GmailAccessTokenExpiry);
        IsSignedIn = false;
        AccountEmail = null;
    }

    // MARK: - PKCE helpers

    private static string RandomCodeVerifier() => Base64UrlEncode(RandomNumberGenerator.GetBytes(32));

    private static string CodeChallenge(string verifier) =>
        Base64UrlEncode(SHA256.HashData(Encoding.UTF8.GetBytes(verifier)));

    private static string Base64UrlEncode(byte[] bytes) =>
        Convert.ToBase64String(bytes).Replace('+', '-').Replace('/', '_').TrimEnd('=');

    /// Extracts the email claim from a JWT id_token without verifying it
    /// (verification isn't needed — the token came directly from Google over TLS).
    private static string? EmailFromIdToken(string idToken)
    {
        var parts = idToken.Split('.');
        if (parts.Length < 2) return null;
        var payload = parts[1].Replace('-', '+').Replace('_', '/');
        payload = payload.PadRight(payload.Length + (4 - payload.Length % 4) % 4, '=');
        try
        {
            var bytes = Convert.FromBase64String(payload);
            using var doc = JsonDocument.Parse(bytes);
            return doc.RootElement.TryGetProperty("email", out var email) ? email.GetString() : null;
        }
        catch (FormatException) { return null; }
        catch (JsonException) { return null; }
    }

    // MARK: - Token model

    private sealed class TokenResponse
    {
        [JsonPropertyName("access_token")] public string? AccessToken { get; set; }
        [JsonPropertyName("refresh_token")] public string? RefreshToken { get; set; }
        [JsonPropertyName("expires_in")] public int? ExpiresIn { get; set; }
        [JsonPropertyName("id_token")] public string? IdToken { get; set; }
    }
}
